### ISSUE TITLE
Login: URL field does not recognize emoji domains

### ISSUE THREAD

**Issue by sandymcfadden** [wordpress-mobile/WordPress-iOS#6317]
### Expected behavior
- Adding a self hosted site which has an emoji in the domain should be a valid URL and allow you to login

### Actual behavior
- When adding a domain with emoji the Add Site button will not highlight so it can be clicked.

### Steps to reproduce the behavior
- Go to My Sites
- Add Self Hosted SIte
- Enter username and password
- Enter a domain with an emoji in it and try to add.
- Puny code for the domain does not validate either.

Reported in #147692-h
Ping: @astralbodies

**Comment by astralbodies** [comment]
@koke It looks like the NSURL+IDN methods are returning nil when the emoji-based URL is being passed in. I can file an issue in that repo - looks like we should probably modernize that project and add some good unit testing.

**Comment by koke** [comment]
I'm tempted to just replace NSURL-IDN with a Swift wrapper for [libidn](https://www.gnu.org/software/libidn/) but I'm really really not in the mood for C right now 😬

**Comment by koke** [comment]
The ideal solution would be to use ICU's [uidna](http://icu-project.org/apiref/icu4c/uidna_8h.html) functions, which is what WebKit uses. Unfortunately the headers are missing from iOS, and this would mean bundling our own version of ICU. I've considered doing that in the past to work around some other i18n issues, so we might as well do it and start leveraging that wonderful library.

**Comment by kwonye** [comment]
Same with the Russian keyboard

**Comment by kwonye** [comment]
@aerych any ideas?

**Comment by aerych** [comment]
@kwonye I think what @koke suggests is the best thing for a permanent fix.  For the mean time, a work around is to enter the [punycode](https://en.wikipedia.org/wiki/Punycode) equivalent instead of the emoji.  Maybe we can refer folks to https://www.punycoder.com/ or something similar.

**Comment by koke** [comment]
Radar created: http://www.openradar.me/29948279
This should really be baked into NSURL 😞

**Comment by aerych** [comment]
> This should really be baked into NSURL

Seriously!

**Comment by astralbodies** [comment]
Maybe something simple like this?

```swift
let urlWithEmoji = "http://hello☢️.com"

var allowedCharacters = CharacterSet.urlPathAllowed.union(.urlHostAllowed)
let urlWithEmojiEscaped = urlWithEmoji.addingPercentEncoding(withAllowedCharacters: allowedCharacters)
let url = URL(string: urlWithEmojiEscaped!)
```

**Comment by astralbodies** [comment]
Ahha .. the percent-encoding problem with NSURLSession...

**Comment by kwonye** [comment]
@aerych this is fixed on our end, correct? Can we close this out?

**Comment by aerych** [comment]
> this is fixed on our end, correct? Can we close this out?

Not really, and no. :)  #6378 did not resolve the issue with (specifically) emoji, only with non-latin characters.

**Comment by koke** [comment]
I've spent some time trying to figure out how to implement IDNA properly and Punycode is not an easy algorithm to digest. The problem is that NSURL-IDN uses some code copied directly from [OmniFoundation](https://github.com/omnigroup/OmniGroup/tree/master/Frameworks/OmniFoundation) and I don't really understand the algorithm enough to fix it. This would require quite some effort to fix for a small impact.

I wasn't sure how to approach this because even if it affected only a small set of users, it felt wrong that they couldn't use the app because we badly implemented a standard. Until I kept reading and arrived to this: [Why can't you register emojis in gTLDs](http://domainincite.com/18035-why-you-cant-register-emojis-in-gtlds):

> Emoticon domains only work in software that has not implemented the latest IDNA standard. Only the older, deprecated version of the IDNA standard allowed emoticons, more or less by accident. Over time, these domains will increasingly not work correctly as software vendors update their implementations.

Looking at the [relevant RFC](https://tools.ietf.org/html/rfc5892#appendix-B) confirms it, and IDNA2008 does not allow emoji on labels. This means I feel much more comfortable with a _wontfix_.

I have to say however, that there's something odd with our Punycode implementation, as [❤️❤️❤️.ws](http://demo.icu-project.org/icu-bin/idnbrowser?t=❤%EF%B8%8F❤%EF%B8%8F❤%EF%B8%8F.ws) should encode to `xn--qeiaa.ws`, but it's encoding to `xn--qeiaa2390tbab.ws`. It makes me really nervous that we have some algorithm that's not behaving correctly and we don't understand what it does nearly enough to change it.
